### DISCUSSION TITLE
RTK and RTK Query ?

### DISCUSSION THREAD

**Discussion by omdxp** [dzcode-io/dzcode.io#455] [Architecture]
[Redux Toolkit](https://redux-toolkit.js.org) is an amazing tool for redux applications, it has useful packages within namely `redux`, `immer` and even `redux-thunk` so as you can see it includes a lot of toolset that are needed to any common projects.

It doesn't have a learning curve like Redux, as a matter of fact it minimize a lot of the boilerplates with normal Redux and enhance the dx side of things. And what I like the most about is the [RTK Query](https://redux-toolkit.js.org/rtk-query/overview) which has an amazing api like [React Query](https://react-query.tanstack.com) but with the bonus of directly integrated with Redux.

So what do you think about it ?

**Comment by ZibanPirate** [comment]
i strongly disagee here, this is in itself another wrapper around Redux:
- New devs won't understand how Redux work, and would take RTK for granted, which defeats the purpose of dzcode.
- Another dependency to maintain and keep up-to-date.
- We will most likely go with [GraphQL](https://github.com/dzcode-io/dzcode.io/discussions/445#discussioncomment-2204422), and keep redux only for Non-fetchable data, such as Settings, this will leave us with very few cases to use Redux.

Once we implement GraphQL, I'm planning to use these three state management tools:

- GraphQL for fetchable data, eg: contribution cards in `/Contribute`.
- Redux for Global Settings, eg: DarkMode, Display Language.
- useState hook for local state, eg: filters in `/Contribute`, where the initial state would be taken from URL query, and when the state changes, it would also change the URL.

So all three are justifiable, and they would be a good learning experience, we can even write an article on how we use them and link live code snippets.

**Reply by omdxp** [reply]
Sounds good to me. Just a fun fact here, RTK is suggested from the Redux team and it's becoming like the standard for Redux since the beginning of 2022, you can check by hovering on `createStore` function exported from `redux` as it is deprecated and they suggest using `configureStore` from `@reduxjs/toolkit` in the newer versions of `redux`. However, as you said, since we are planning to use GraphQL, a lot of the cases in the redux code will be eliminated so I'm down for that! 🙏🏻

**Reply by ZibanPirate** [reply]
> RTK is suggested from the Redux team and it's becoming like the standard for Redux since the beginning of 2022

I guess I jumped to a conclusion without checking the sate of Redux now, it's been a while since we set up Redux on dzcode.io, and things have changed since then, if we want to upgrade our dependencies, then we probably need to use Redux Toolkit.

Still, I don't think we have a need for RTK Query, what do you think?

**Reply by omdxp** [reply]
It'll be a huge jump as it helps out with a lot of use cases like normalizing state automatically (not like the old days when you have to do it manually), so I believe it is a good idea to start using it.

About RTK Query, it is included in the `@reduxjs/toolkit` package and from my experience it would help a lot with API calls, and it is integrated directly to Redux store and it has a lot of helpful utilities like invalidate some calls based on defined tags and create custom hooks on the fly to use them directly. However, in our case we are just querying data and not mutating, so I think we can stay with our `fetchV2` since we are just getting data.
